### DISCUSSION TITLE
[4주차] 트랜스포트 계층 3.5~8

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Dijkstra-Employ/Network-Study#12] [Q&A]
4주차 트랜스포트 계층 3.5~8에 해당하는 질의응답입니다.

**Comment by H-sooyeon** [comment]
## Q1

각 TCP 연결에는 Receive Buffer와 Send Buffer가 존재합니다.
이는 송신측과 수신측 각각 가지고 있는 버퍼이며, 이를 통해 데이터를 전달하고 받습니다.
송신측과 수신측의 Receive Buffer와 Send Buffer 데이터 전송 흐름을 cwnd를 활용하여 간단하게 설명해주세요.

## Q2

TCP 혼잡 제어 정책 중, 3 ACK Duplicated와 Timeout에 대한 처리 방식이 다릅니다.
이는 네트워크 혼잡 정도의 차이를 판단하여 처리한 것입니다.
각각의 경우에 대한 처리 과정의 차이, TCP Tahoe와 TCP Reno의 동작 과정에 대해 설명해주세요.

**Reply by choiyoung69** [reply]
### Q1
전송하려는 데이터가 먼저 Sendbuffer에 저장된다. cwnd는 송신 측이 네트워크의 혼잡 상태를 감지하여 한 번에 전송할 수 있는 송신자의 데이터 양을 제한한다. cwnd는 데이터가 성공적으로 도달하게 될 때 cwnd가 늘어나고, 손실이 일어나면 cwnd가 발생한다. 그리고, 송신 측은 cwnd와 rwnd 중 작은 값을 기준으로 데이터를 전송한다. 
수신 측에 도착한 데이터는 먼저 Receive Buffer에 저장된다. 수신한 데이터를 애플리케이션에 전달하며, 동시에 송신측에 ACK를 전송한다. 이떄, 수신 측은 자신의 Receive Buffer의 의 남은 공간 크기(rwnd)를 송신측에 알려준다. 

---

### Q2
**3 ACK Duplciated** 
3개의 중복된 ACK는 패킷이 손실되었지만, 그 후의 데이터는 안전하다는 의미로 심각하게 혼잡하지는 않다.

- TCH Tahoe : 3개 중복 ACK를 수신하면 Fast Retransmit을 실행, 손실된 패킷을 즉시 재전송한다. 이와 동시에 Slow Start로 돌아가며, cwnd를 1 MSS로 줄입니다. 이후 Slow Start 과정을 통해 cwnd를 증가시킨다. 이 과정에서 Slow Start Threshold를 cwnd의 절반으로 설정하고 이후에 Congestion Avoidance 알고리즘이 적용된다.

- TCP Reno:
TCP Reno도 Fast Retransmit을 실행하여 패킷을 재전송한다. 그러나, Fast Recovery를 통해 성능을 최적화하고, ssthresh를 cwnd의 절반으로 설정하고 cwnd를 ssthresh로 줄인 후, Congestion Avoidance 모드로 전환한다. 이떄 ACK가 올 때마다 cwnd를 1MSS씩 증가시킨다. ssthresh(Slow Start Threshold)를 현재 cwnd의 절반으로 설정합니다. 이후에는 Congestion Avoidance 알고리즘이 적용


**Timeout** 
발생시 패킷이 손실되었고, 손실된 패킷에 대해 송신측이 ACK를 받지 못했다는 것을 의미한다. 즉, 네트워크가 심각하게 혼잡하다는 것을 의미한다. 

- TCP Tahoe : 타임아웃이 발생하면 TCH Tahoe는 Slow Start 상태로 돌아가서, cwnd를 1 MSS로 줄이고, sstresh를 cwnd의 절반으로 줄인다. 이후, 처음부터 다시 cwnd를 증가시킨다.

- TCH Reno : 타임아웃 발생 시 Slow Start로 돌아가고, cwnd를 1 MSS로 줄인 후 ssthresh를 cwnd의 절반으로 줄인다. 이후 TCH Tahoe와 동일하게 SLow Start 과정을 통해 cwnd를 증가시킨다.

**Reply by Hanjaemo** [reply]
> ### Q1
> 각 TCP 연결에는 Receive Buffer와 Send Buffer가 존재합니다.
> 이는 송신측과 수신측 각각 가지고 있는 버퍼이며, 이를 통해 데이터를 전달하고 받습니다.
> 송신측과 수신측의 Receive Buffer와 Send Buffer 데이터 전송 흐름을 cwnd를 활용하여 간단하게 설명해주세요.

1. 송신 측에서 전송하려는 데이터를 Send Buffer에 저장합니다.
2. TCP는 Send Buffer에 있는 데이터들 중에서 송신 측 cwnd와 수신 측 rwnd 최솟값만큼의 데이터만 세그먼트에 캡슐화하여 전송합니다.
3. 수신 측에서는 세그먼트를 성공적으로 수신하여 Receive Buffer에 저장합니다. 
수신 측 애플리케이션이 Receive Buffer에 있는 데이터를 읽어가면서 Receive Buffer에 여유 공간이 생기면, 수신 측 TCP는 ACK 세그먼트에 rwnd를 담아 전달하여 송신 측에 이를 알립니다.
4. 송신 측에서는 수신된 ACK 세그먼트에 담긴 rwnd와 cwnd를 참고하여 데이터 전송량을 조절합니다.

> ### Q2
> TCP 혼잡 제어 정책 중, 3 ACK Duplicated와 Timeout에 대한 처리 방식이 다릅니다.
> 이는 네트워크 혼잡 정도의 차이를 판단하여 처리한 것입니다.
> 각각의 경우에 대한 처리 과정의 차이, TCP Tahoe와 TCP Reno의 동작 과정에 대해 설명해주세요.

3 ACK Duplicated나 Timeout이 발생했을 때, cwnd 크기가 감소하지만 다음과 같은 차이가 존재합니다.

- 3 ACK Duplicated의 경우, timeout에 비해 혼잡 정도가 크지 않기 때문에 굳이 많은 손해를 보면서까지 슬로 스타트로 돌아갈 필요가 없습니다. 따라서 빠른 회복 단계로 전환되어 cwnd 크기를 절반으로 감소시킨 다음 곧바로 손실된 세그먼트를 재전송합니다. 그리고나서 슬로 혼잡 회피 단계로 전환하여 cwnd를 선형적으로 증가시킵니다.
- Timeout의 경우, 전송한 패킷 뿐만 아니라 ACK까지 송신 측에 도착하지 않은 상태를 의미합니다. 이때 송신 측에서는 네트워크 혼잡 정도가 심각하다고 판단하여, cwnd를 1MSS로 대폭 줄인 다음 슬로 스타트 단계로 돌아갑니다.

틀린 답변이 있을 경우 코멘트 남겨주세요!!!

**Reply by Starlight258** [reply]
> ## Q1
> 각 TCP 연결에는 Receive Buffer와 Send Buffer가 존재합니다. 이는 송신측과 수신측 각각 가지고 있는 버퍼이며, 이를 통해 데이터를 전달하고 받습니다. 송신측과 수신측의 Receive Buffer와 Send Buffer 데이터 전송 흐름을 cwnd를 활용하여 간단하게 설명해주세요.
> 
0. 송신측에서 보낼 데이터를 `송신 버퍼`에 저장한다.
1. 송신측에서 `min{cwnd, rwnd}` 데이터를 송신 버퍼에서 가져와 전송한다.
2. 수신측은 받은 데이터를 `수신 버퍼`에 저장하고, `ACK`를 `rwnd`와 함께 송신측에 보낸다.
3. 송신측은 수신한 `ACK`를 `수신 버퍼`에 저장하고 `cwnd`와 `rwnd`를 고려하여 데이터 전송량을 조절한다.

> ## Q2
> TCP 혼잡 제어 정책 중, 3 ACK Duplicated와 Timeout에 대한 처리 방식이 다릅니다. 이는 네트워크 혼잡 정도의 차이를 판단하여 처리한 것입니다. 각각의 경우에 대한 처리 과정의 차이, TCP Tahoe와 TCP Reno의 동작 과정에 대해 설명해주세요.

- 3중 중복 ACK : 
  - TCP Tahoe : `sshtresh`를 `cwnd/2`로 설정하고, `cwnd`는` 1`로 설정한 후 슬로 스타트를 수행한다.
  - TCP Reno : `ssthresh`를 `cwnd/2`로 설정하고, `cwnd`는 `ssthresh + 3 MSS`로 설정한 뒤 빠른 회복 상태로 들어간다.

- 타임아웃:
  -  TCP Tahoe와 TCP Reno 모두 `sshtresh`를 `cwnd/2`로 설정하고, `cwnd`는 `1`로 설정한 후 슬로 스타트를 수행한다.

**Comment by Starlight258** [comment]
## Q1
TCP 세그먼트 크기에 대해 `MSS`와 `MTU` 키워드를 포함해 설명해주세요.

## Q2
TCP 혼잡 제어 알고리즘에서, `타임아웃`과 `3중 중복 ACK`를 다르게 처리하는 이유가 무엇일까요?
각 이벤트는 네트워크의 어떤 상태를 의미하나요?

**Reply by H-sooyeon** [reply]
> Q1 TCP 세그먼트 크기에 대해 MSS와 MTU 키워드를 포함해 설명해주세요.

TCP 세그먼트 크기와 관련해 MSS와 MTU은 네트워크에서 데이터를 효율적으로 전송하고 패킷 손실을 최소화하는 데 중요한 역할을 한다.
MSS는 TCP 세그먼트에서 전송할 수 있는 최대 데이터 크기를 나타내며, IP 헤더와 TCP 헤더를 제외한 순수 데이터의 크기를 의미한다.
MTU는 네트워크 인터페이스가 한 번에 전송할 수 있는 최대 패킷 크기를 의미하며, 헤더를 포함한 전체 패킷의 크기를 나타낸다.
만약 TCP 세그먼트가 MTU보다 크다면, 네트워크 장치는 이를 작은 조각으로 나누어 전송해야 하며, 이 과정에서 지연이나 패킷 손실이 발생할 수 있다. 따라서 TCP는 MSS를 기준으로 세그먼트 크기를 조정하여 단편화를 최소화해야 한다.

<br />

> Q2 TCP 혼잡 제어 알고리즘에서, 타임아웃과 3중 중복 ACK를 다르게 처리하는 이유가 무엇일까요?
각 이벤트는 네트워크의 어떤 상태를 의미하나요?

다르게 처리하는 이유는 네트워크의 혼잡 상태가 같지 않다고 판단하기 때문이다.
타임아웃은 심각한 혼잡이나 네트워크 장애로 판단하여 전송 속도를 크게 줄이며, 3중 중복 ACK는 경미한 혼잡이나 부분 패킷 손실로 판단하여 빠른 재전송과 혼잡 회피 알고리즘을 사용해 전송 속도를 조절한다.

**Reply by choiyoung69** [reply]
### Q1
세그먼트로 담을 수 있는 최대 데이터의 양은 최대 세그먼트 크기(MSS)로 제한된다.순수 데이터) 그리고, MSS는 TCP/IP 헤더를 포함하고 있는 가장 큰 링크 계층 프레임의 길이(MTU)에 의해 결정된다.

---

### Q2
두 이벤트가 네트워크의 서로 다른 상태를 나타내기 떄문이다. 따라서, 각각 이벤트가 발생할 때 상황을 다르게 해석하고 상황에 맞는 적절한 조치를 취해주는 것이다.

- Timeout : 일반적으로 네트워크가 심각하게 혼잡하거나, 패킷이 아예 도달하지 못했을 가능성이 큰 상항을 의미한다.
- 3중 중복 ACK : 네트워크는 여전히 동작중이며, 어느 정도의 경미한 혼잡이 있거나, 일부 패킷만 손실이 발생한 상황이다.

**Reply by Hanjaemo** [reply]
> ## Q1
> TCP 세그먼트 크기에 대해 `MSS`와 `MTU` 키워드를 포함해 설명해주세요.

MSS(Maximum Segment Size)란  TCP 세그먼트에 담을 수 있는 최대 데이터 크기를 의미합니다. TCP는 이 단위로 데이터를 나누어 전송합니다.

그리고 MTU(Maximum Transmission Unit)는 네트워크에서 전송할 수 있는 가장 큰 프레임의 크기를 의미합니다. 각 네트워크 요소는 자체적으로 MTU 값을 가지며, 이 값에 따라 패킷의 크기가 결정됩니다.

예를 들어, 링크 계층의 요소가 2000바이트의 MTU를 가지고 있을 경우, 상위 계층에서는 이 값보다 큰 패킷을 링크 계층에 전달할 수 없습니다.

이더넷과 PPP 링크 계층 프로토콜은 모두 1500바이트의 MTU를 가지고 있어, 일반적으로 MSS는 약 1460바이트로 정의된다고 합니다. (`MSS = MTU - IP 헤더 크기 - TCP 헤더 크기 = 1500 - 20 - 20`)

> ## Q2
> TCP 혼잡 제어 알고리즘에서, `타임아웃`과 `3중 중복 ACK`를 다르게 처리하는 이유가 무엇일까요? 각 이벤트는 네트워크의 어떤 상태를 의미하나요?

3 ACK Duplicated나 Timeout이 발생했을 때, cwnd 크기가 감소하지만 다음과 같은 차이가 존재합니다.

- 3 ACK Duplicated의 경우, timeout에 비해 혼잡 정도가 크지 않기 때문에 굳이 많은 손해를 보면서까지 슬로 스타트로 돌아갈 필요가 없습니다. 따라서 빠른 회복 단계로 전환되어 cwnd 크기를 절반으로 감소시킨 다음 곧바로 손실된 세그먼트를 재전송합니다. 그리고나서 슬로 혼잡 회피 단계로 전환하여 cwnd를 선형적으로 증가시킵니다.
- Timeout의 경우, 전송한 패킷 뿐만 아니라 ACK까지 송신 측에 도착하지 않은 상태를 의미합니다. 이때 송신 측에서는 네트워크 혼잡 정도가 심각하다고 판단하여, cwnd를 1MSS로 대폭 줄인 다음 슬로 스타트 단계로 돌아갑니다.

틀린 답변이 있을 경우 코멘트 남겨주세요!!!

**Reply by Starlight258** [reply]
> ## Q1
> TCP 세그먼트 크기에 대해 `MSS`와 `MTU` 키워드를 포함해 설명해주세요.
> 
TCP는 데이터를 최대 세그먼트 크기인 `MSS` 이하로 나눈다.

`MSS`는 헤더를 포함하지 않은 **애플리케이션 계층 데이터에 대한 최대 크기**이다.
`MTU`는 **네트워크 링크에서 전송될 수 있는 가장 큰 패킷의 크기**이다. (IP 헤더와 TCP 헤더를 포함한다.)

> `MSS` = `MTU` - `IP 헤더 크기` - `TCP 헤더 크기`

> ## Q2
> TCP 혼잡 제어 알고리즘에서, `타임아웃`과 `3중 중복 ACK`를 다르게 처리하는 이유가 무엇일까요? 각 이벤트는 네트워크의 어떤 상태를 의미하나요?

- `타임아웃` : 혼잡 이벤트
	- 네트워크의 **심각한 혼잡상태**를 의미하며 공격적인 대응을 한다.
- `3중 중복 ACK` : 손실 이벤트
	- 네트워크의 **일시적인 혼잡상태**를 의미하며 온건한 대응을 한다.

**Comment by Hanjaemo** [comment]
### Q1
TCP가 혼잡 제어를 위해 접근하는 2가지 방식에 대해 설명해 주세요.

### Q2
혼잡 제어에 필요한 네트워크 혼잡 임곗값은 어떤 기준으로 지정하나요?

**Reply by H-sooyeon** [reply]
> Q1 TCP가 혼잡 제어를 위해 접근하는 2가지 방식에 대해 설명해 주세요.

혼잡 제어에 대한 TCP의 접근법은 두 가지가 있다.
1. 종단간의 혼잡제어
2. 네트워크 지원 혼잡제어

종단간의 혼잡제어는 네트워크 중간장치가 혼잡에 대한 지원을 해주지 않기에 송신과 수신 종단 TCP에서 혼잡 상황을 추론해야 한다.
이때 Timeout, 3 Duplicated ACK 발생 시 혼잡이 발생한 것으로 간주하여 window size를 줄이는 등의 혼잡 제어를 실시한다.

네트워크 지원 혼잡 제어는 네트워크 계층에서 네트워크의 혼잡 상태를 판단하여 종단 시스템에게 직접적인 피드백을 제공한다.
IP 데이터그램 헤더의 서비스 유형 필드에 있는 2비트가 ECN에 사용되어 라우터가 정체를 겪고 있음을 TCP 송-수신자에게 알려준다.

<br />

> Q2 혼잡 제어에 필요한 네트워크 혼잡 임곗값은 어떤 기준으로 지정하나요?

ssthresh는 네트워크 상태와 패킷 손실 이벤트에 따라 동적으로 설정된다.
초기 설정시엔 매우 큰 값으로 설정되며, 이후 패킷 Loss 및 ssthresh 재설정 등의 과정을 거쳐 재조정을 한다.
일반적으로, ssthresh는 패킷 Loss 발생 시 현재 혼잡 윈도우 크기의 절반으로 설정된다.
TCP Reno에선 타임아웃 발생 시 cwnd는 1MSS로 줄어들지만 ssthresh는 그대로 유지된다.

**Reply by choiyoung69** [reply]
### Q1
- 종단 간 혼잡 제어 : 중간의 네트워크 장비들에게 직접적인 정보를 받지 않고 종단에서 네트워크 상태를 추론하여 혼잡 제어를 수행하는 것이다. TImeout이나 3중 중복 ACK 발생 시  혼잡 제어를 실시하는 등의 조치를 취한다.
- 네트워크 지원 혼잡 제어 : 라우터가 네트워크의 혼잡상태를 감지하고 이를 송수신 측에 알리는 방식이다. IP 데이터그램 헤드에 있는 2비트 ECN이 송수신자에게 혼잡을 알리는데 사용된다. 

---

### Q2
처음에 혼잡 임계값 ssthresh는 일반적으로 매우 큰 값(최대 윈도우 크 등)으로 설정된다. 그러나 혼잡이 감지되면 ssthresh는 보통 cwnd의 절반으로 설정되고, Slow Start와 Congestion Avoidance 간의 전환을 결정하는 중요한 기준이 된다.

**Reply by Starlight258** [reply]
> ### Q1
> TCP가 혼잡 제어를 위해 접근하는 2가지 방식에 대해 설명해 주세요.
> 
1. `종단간의 혼잡 제어`
- 네트워크 계층은 전송 계층에게 **혼잡 제어를 위한 정보를 제공하지 않는다.**
- `타임아웃` 또는 `3중 중복 ACK`가 나타날 때 네트워크 혼잡으로 인해 세그먼트 손실이 일어났다고 유추한다.
- `증가하는 RTT`를 네트워크 혼잡 증가를 나타내는 것으로 유추한다.

2.` 네트워크 지원 혼잡 제어`
- 네트워크 계층 구성요소는 네트워크 안에서 혼잡 상태와 관련해서 **송신자에게 직접적인 피드백을 제공**한다.
- 네트워크 `라우터`에서 송신자에게 직접 보내거나(직접 피드백), 송신자->수신자 `패킷 안의 특정 필드`에 표시한다(왕복 시간만큼 지연 발생).

> ### Q2
> 혼잡 제어에 필요한 네트워크 혼잡 임곗값은 어떤 기준으로 지정하나요?
- 초기 `ssthresh`(slow start threshold)는 매우 높게 설정되며, **패킷 손실**이 감지되거나(3 중복 ACK) **타임아웃** 발생시 `cwnd`의 절반으로 설정한다.

**Comment by choiyoung69** [comment]
## Q1
TCP 세그먼트 헤더 중에는 순서 번호 필드와 확인 응답 번호 필드가 있습니다. 두 필드를 각각 어떤 값으로 채우는지 설명해주세요.

## Q2
혼잡 회피로 전환되는 조건은 무엇인가요?

**Reply by H-sooyeon** [reply]
> Q1 TCP 세그먼트 헤더 중에는 순서 번호 필드와 확인 응답 번호 필드가 있습니다. 두 필드를 각각 어떤 값으로 채우는지 설명해주세요.

순서 번호는 송신 측에서 보낸 바이트 스트림의 시작 지점을 나타낸다.
TCP 연결이 처음 설정될 때, 송신 측은 임의의 초기 순서 번호를 선택한다. 이후 순서 번호 필드는 해당 세그먼트에서 처음 전송되는 바이트의 위치를 표시한다.

ex > ISN이 1000이고 100바이트의 데이터가 첫 번째 세그먼트라면, 첫 번째 세그먼트의 순서 번호 필드는 1000이 된다.
두 번째 세그먼트의 순서 번호는 1100이 된다.

<br />

> Q2 혼잡 회피로 전환되는 조건은 무엇인가요?

혼잡 회피로 전환되는 조건은 네트워크 혼잡이 발생하거나 발생할 가능성이 있다고 판단될 때이다.
이는 주로 두 가지 상황에서 전화되는데,
1. timeout
2. 3 Duplicated ACK

혼잡 회피 단계로 전환되면, TCP는 네트워크 혼잡을 완화하기 위해 전송 속도를 선형적으로 증가시키며 네트워크의 용량을 신중하게 활용하려 한다.

**Reply by Hanjaemo** [reply]
> ## Q1
> TCP 세그먼트 헤더 중에는 순서 번호 필드와 확인 응답 번호 필드가 있습니다. 두 필드를 각각 어떤 값으로 채우는지 설명해주세요.
- **순서 번호 필드**: 송신 측에서 전송하는 데이터 스트림의 첫 번째 위치를 나타내는 값으로 채웁니다.
- **확인 응답 번호 필드**: 수신 측에서 다음에 받아야 할 데이터의 순서 번호로 채웁니다.

> ## Q2
> 혼잡 회피로 전환되는 조건은 무엇인가요?

혼잡 회피의 목적은 네트워크 혼잡을 최소화하면서도 가능한 최대 전송 속도를 유지하는 것입니다.

이러한 혼잡 회피로 전환되는 시점은 cwnd 크기가 혼잡 임곗값에 도달할 때입니다.

**Reply by Starlight258** [reply]
> ## Q1
> TCP 세그먼트 헤더 중에는 순서 번호 필드와 확인 응답 번호 필드가 있습니다. 두 필드를 각각 어떤 값으로 채우는지 설명해주세요.
> 
 - 순서 번호 필드: 세그먼트의 첫 번째 바이트가 전체 스트림에서 차지하는 위치
 - 확인 응답 번호 : 호스트 A의 ACK는 호스트 A가 호스트 B로부터 기대하는 다음 바이트의 순서번호

> ## Q2
> 혼잡 회피로 전환되는 조건은 무엇인가요?
- 슬로 스타트: `cwnd`가 `ssthresh`와 같으면, **슬로 스타트는 종료**되고 TCP는 **혼잡 회피 모드**로 전환한다.
- 빠른 회복 : **새로운 ACK** 수신시 `cwnd`를 `ssthresh`로 설정하고 **혼잡 회피**로 전환한다.
